OPHD/XmlSerializer.h

```diff
@@ -1,9 +1,49 @@
 #pragma once
 
+#include <NAS2D/StringUtils.h>
 #include <NAS2D/Xml/Xml.h>
 #include <string>
+#include <unordered_map>
+#include <typeinfo>
+#include <type_traits>
+#include <stdexcept>
 
 
 // Open an xml document
 // Throws a runtime_error if the xml is ill formed or the root element name is incorrect
 NAS2D::Xml::XmlDocument openXmlFile(std::string filename, std::string rootElementName);
+
+int elementToInt(const NAS2D::Xml::XmlElement* element);
+
+template<typename T>
+T stringToEnum(const std::unordered_map<std::string, T>& table, std::string value)
+{
+	auto it = table.find(NAS2D::toLowercase(value));
+	if (it != table.end()) {
+		return it->second;
+	}
+
+	throw std::runtime_error("Unable to parse enum with value of " + value);
+}
+
+
+// Create an overload to parse specific enums
+template<typename T, std::enable_if_t<!std::is_enum<T>::value, bool> = true>
+void parseElementValue(T& destination, const NAS2D::Xml::XmlElement* element)
+{
+	if constexpr (std::is_same_v<T, int>) {
+		destination = elementToInt(element);
+	}
+	else if constexpr (std::is_same_v<T, float>) {
+		destination = static_cast<float>(std::atof(element->getText().c_str()));
+	}
+	else if constexpr (std::is_same_v<T, double>) {
+		destination = std::atof(element->getText().c_str());
+	}
+	else if constexpr (std::is_same_v<T, std::string>) {
+		destination = element->getText();
+	}
+	else {
+		throw std::logic_error("Unable to parse type " + std::string(typeid(T).name()) + " from XML");
```

Comment:
I didn't realize you could disable run time type information. 

Procedures: https://docs.microsoft.com/en-us/cpp/build/reference/gr-enable-run-time-type-information?view=msvc-160

I tried disabling run time type information and received about a 7 warnings: 

> Warning	C4541	'dynamic_cast' used on polymorphic type 'Control' with /GR-; unpredictable behavior may result	\OPHD\UI\Core\RadioButton.cpp	67	

The program crashes when this dynamic_cast is reached, which is before the typeid call mentioned by @DanRStevens, so I wasn't able to directly test. Oddly, it didn't provide a warning about the typeid like the dynamic_cast did. 

I guess we probably do not want to remove the concept of dynamic_cast so probably safe to use typeid? Although it is good to know these 2 features are tied to a compiler settings. I had just assumed they were always available if you had a c++11 compliant compiler since they are features of the language itself... C++ seems to constantly find the most complex ways to express itself.

Reply:
That was an interesting investigation. I had kind of forgotten that `dynamic_cast` was also related to RTTI. I used to avoid it a bit for that reason. It is useful though, and I can see us continuing to use it in this project, so probably safe to assume that RTTI is enabled.